### ISSUE TITLE
Feature Request: Display A.S.SP on site 

### ISSUE THREAD

**Issue by JanelleJenstad** [projectLEMDO/lemdoIssues#7]
Requested by: Janelle Jenstad and the Interface Team at MRU (Olivia Brodowski, Madeline Scarlett, and Cliff Werier)

### What do you want to be able to do or what functionality do you want to see?
We want to view A.S.SP numbers for plays on the webpages of the modern texts (i.e., files with LEMDO Editorial Treatment letModernSpelling).

### Do we have encoding in place? 
Yes

### If yes, what is the standard encoding?
Speeches are numbered continuously throughout a scene. Numbering restarts at 1 with each new scene. The speech number can be derived from the last part of the xml:id on the <sp> element, which will always be _sp + a number. 

For plays with scenes only (example from FV):
`<sp who="#emdFV_M_Henry" xml:id="emdFV_M_s1_sp1">`
For plays with acts and scenes (example from H5):
`<sp who="#emdH5_FM_Canterbury" xml:id="emdH5_FM_a1_s1_sp1">`

### If not, what do you need added to or allowed by the schema? [Or how do you want to encode something? Give an example that is allowed by the TEI.]
n/a

### What should happen to this markup in webpages?
What to display:
o	1.1.Sp1 [for plays like H5 that have acts and scenes]
o	1.Sp1 [for plays like FV that have only scenes]

### How to display it:
o	In a sans serif font (Alegreya Sans)
o	Same size as the play text font (with the intention that the line numbers from PDFs that we later add into the digital editions later be smaller and be rendered in gray, as is the convention for line numbers).
o	On the left side of the text block, quite close to the text block. [The more white space between the text block and the number, the harder it is for users to figure out what number they need.]
o	Display all speech numbers, even in cases of rapid-fire dialogue where speeches might be one line long.
o	If the line is split between speakers, we will still display all speech numbers. [I.e., in these cases, speech numbers will actually be more granular than line numbers.]

### Clicking on it should generate this behaviour:
o	Page aligns to the top of the speech
o	URL in the browser bar is [[general path]]/filename#xml:id

Example for FV in lemdo-dev:
https://jenkins.hcmc.uvic.ca/job/LEMDO/lastSuccessfulBuild/artifact/products/lemdo-dev/site/emdFV_M.html#emdFV_M_s1_sp1

### Thinking ahead: 
When we also have line numbers (which we don’t yet have and won’t have until the printed book is published), we want to be able to toggle between line numbers and speech numbers by clicking a radio button in one of the hamburger menu windows. I.e., we can’t do anything now that would preclude toggling later. 

### Can you describe what you think might have to happen in the processing? 
Janelle doesn’t think that we have any processing in place that we can tweak or repurpose. However, once this job is done, we’ll have some baseline processing that we can repurpose to display continuous speech numbers in OS texts, witness line numbers in OS texts, and analeptic line numbers in the modern texts (i.e., the line numbers generated for the PDFs that we’ll add back in as milestones in the digital text).

### What should happen to this markup in PDFs? [If this markup is used in documents that will not be included in the PDFs, write n/a.]
We would like to include speech numbers in the printed text as well as the LaTeX-generated line numbers.

We suggest:

- Speech numbers on the left – just the number (the 3 from xml:id="emdH5_M_a1_s2_sp3) so that they don’t take up too much space
- Line numbers on the right (as they now are). 
- To explain this system, we suggest a bit of boilerplate text that gets auto-included at the bottom of the page where the character list appears or on the next page … a “How to read this playbook” section??

### Where should the documentation go?
We should have some kind of documentation since this is new processing – presumably in the Programming chapter. 
Nicole and Janelle will some interstitial information to our existing documentation about how the encoding will be rendered.

**Comment by martindholmes** [comment]
@JanelleJenstad Do we have any QME texts that are going to be published in print? I'm a bit concerned that there's not enough space in the left margin for speech numbers on the print layout, so I think there'll be major ramifications to deal wit there. I think I would prefer to put the speech number after the speech prefix in parentheses, since there's always room there:

~~~~~
KING HENRY (sp. 27)
  May I with right and conscience make this claim?
~~~~~

rather than end up doing something like this:

~~~~~
KING HENRY
sp. 27  May I with right and conscience make this claim?
~~~~~

which will either push that line over to the right a lot, or push the whole speech over if we want to keep the lines lined up.

**Comment by martindholmes** [comment]
@JanelleJenstad Turning to the web view, we need to slot the info into e.g. this layout:

```
Ely
But how, my lord, shall we resist it now?
Canterbury
It must be thought on. If it pass against us,
We lose the better half of our possession,
```

giving us something like this:

```
          Ely
1.1.Sp2   But how, my lord, shall we resist it now?
          Canterbury
1.1.Sp3  It must be thought on. If it pass against us,
          We lose the better half of our possession,
```

One issue is that using Allegreya, we have a variable-width font, so we need to add lots of padding to be sure that the higher counts are still well-separated from the text itself. This is particularly problematic with e.g. Epilogues, where we either have to adopt a convention and explain it:

```
        Chorus
Ps1.Sp1 Thus far, with rough and all-unable pen,
        Our bending author hath pursued the story,
```

or we have to leave much bigger margins:

```
              Chorus
Epilogue1.Sp1 Thus far, with rough and all-unable pen,
              Our bending author hath pursued the story,
```

In the latter case, we would have to decide whether prologues, epilogues etc. are OK with larger left margins than regular scenes, or whether the entirety of the spoken text would have to be indented the width of the largest label to make things look regular. What do you think?

**Comment by JanelleJenstad** [comment]
On [the comment about QME](https://github.com/projectLEMDO/lemdoIssues/issues/7#issuecomment-1072748893):

Yes, the QME editions will be published in the digital anthology and in print. 

I really like this option:
```
KING HENRY (sp. 27)
  May I with right and conscience make this claim?
```

That's an ingenious solution!

**Comment by JanelleJenstad** [comment]
On the [web view comment](https://github.com/projectLEMDO/lemdoIssues/issues/7#issuecomment-1072755050):

Quick thought about Epilogue -- we could abbreviate to Epi. for the marginal number as well as for the PDFs. 

Another thought is to omit all but the speech number from that column. The header at the top of the page will give scene (or act-scene) or epilogue or prologue.

**Comment by martindholmes** [comment]
Decision from meeting 2022-03-18:

We initially show only the speech number (Sp1) but when you mouse over, the title attribute gives you the full expanded version. This solves the problem for desktop, and avoids the issue with prologues and epilogues. For mobile, we need to also provide a little toggle that will display the full version with a click. This could be done by providing a + character before the link itself, which would just toggle a class on the element to show/hide some hidden content.

**Comment by martindholmes** [comment]
I've just committed an implementation of this for the HTML rendering; let's see what breaks. :-)

**Comment by martindholmes** [comment]
@JanelleJenstad On the issue of the PDFs, I think it would be a mistake to put the speech numbers in the left margin. In order to do that, the entirety of the text needs to be shifted to the right, which will dramatically increase the page count. Instead, I suggest that we put the abbreviated form (1.1.Sp1) in the right margin where the line numbers already go, but next to the speech prefix; since speech prefixes aren't counted as part of the line-count, that line in the right margin is guaranteed to be empty. That should ensure that only empty space is used, so the page-count doesn't rise, and since speech prefixes are typically shorter than regular lines, we should never run into overwrite collisions.

**Comment by JanelleJenstad** [comment]
Completely agree! Great idea.

**Comment by martindholmes** [comment]
I'll go ahead with that, then. Any deadline for the PDF changes?

**Comment by JanelleJenstad** [comment]
End of May at the latest. We want James to be proofreading his H5 text by then so that we can publish it this summer.

**Comment by JanelleJenstad** [comment]
I'm closing this ticket and making a new ticket for PDF changes. Thanks for your good work on the online ASSP numbers!
